core/src/main/java/com/scalar/db/transaction/consensuscommit/PrepareMutationComposer.java

```diff
@@ -137,41 +141,20 @@ private void add(Get base) {
     mutations.add(put);
   }
 
-  private List<Value<?>> createBeforeValues(Mutation base, TransactionResult result) {
+  private List<Value<?>> createBeforeValues(Mutation base, TransactionResult result)
+      throws ExecutionException {
     List<Value<?>> values = new ArrayList<>();
-    result
-        .getValues()
-        .values()
-        .forEach(
-            v -> {
-              if (isBeforeRequired(v, base.getPartitionKey(), base.getClusteringKey())) {
-                values.add(v.copyWith(Attribute.BEFORE_PREFIX + v.getName()));
-              }
-            });
-    return values;
-  }
-
-  private boolean isBeforeRequired(Value<?> value, Key primary, Optional<Key> clustering) {
-    return !value.getName().startsWith(Attribute.BEFORE_PREFIX)
-        && !isValueInKeys(value, primary, clustering);
-  }
-
-  private boolean isValueInKeys(Value<?> value, Key primary, Optional<Key> clustering) {
-    for (Value<?> v : primary) {
-      if (v.equals(value)) {
-        return true;
+    for (Value<?> value : result.getValues().values()) {
+      if (isBeforeRequired(base, value.getName())) {
+        values.add(value.copyWith(Attribute.BEFORE_PREFIX + value.getName()));
       }
     }
+    return values;
+  }
 
-    if (!clustering.isPresent()) {
-      return false;
-    }
-
-    for (Value<?> v : clustering.get()) {
-      if (v.equals(value)) {
-        return true;
-      }
-    }
-    return false;
+  private boolean isBeforeRequired(Mutation base, String columnName) throws ExecutionException {
+    TransactionTableMetadata metadata = tableMetadataManager.getTransactionTableMetadata(base);
+    return !metadata.getPrimaryKeyColumnNames().contains(columnName)
+        && metadata.getAfterImageColumnNames().contains(columnName);
```

Comment:
When the column is not primary key column and is an after image column, update the before column of the column.

With this change, users can create `before_` prefixed columns as normal columns. If the column name is `before_x`, the before column name is `before_before_x`.

Reply:
Why don't you use just `com.scalar.db.transaction.consensuscommit.TransactionTableMetadata#getBeforeImageColumnNames()` ?

Reply:
But `return !metadata.getBeforeImageColumnNames().contains(columnName)` contains the primary key columns? Or we can use `return !metadata.getPrimaryKeyColumnNames().contains(columnName) && !metadata.getBeforeImageColumnNames().contains(columnName)`, but it looks like the current one is more straightforward. What do you think?

Reply:
Oops, got it. I had a wrong idea, cluster key has `before` column while partition key doesn't have. Please ignore the previous comment.